src/modules/esl-popup/core/esl-popup-position.ts

```diff
@@ -0,0 +1,196 @@
+import {Rect} from '../../esl-utils/dom/rect';
+
+export type PositionType = 'top' | 'bottom' | 'left' | 'right';
+
+export interface ArrowPositionValue {
+  x: number;
+  y: number;
+  position: string;
+}
+
+export interface PopupPositionValue {
+  x: number;
+  y: number;
+  arrow: ArrowPositionValue;
+}
+
+export interface IntersectionRatioRect {
+  top?: number;
+  left?: number;
+  right?: number;
+  bottom?: number;
+}
+
+export interface PopupPositionConfig {
+  position: PositionType;
+  behavior: string;
+  intersectionRatio: IntersectionRatioRect,
+  element: DOMRect;
+  inner: Rect;
+  outer: Rect;
+  trigger: Rect;
+}
+
+/**
+ * Calculate Rect for given popup position config.
+ * @param cfg - popup position config
+ * */
+function calcPopupBasicRect(cfg: PopupPositionConfig): Rect {
+  let x = cfg.inner.cx - cfg.element.width / 2;
+  let y = cfg.inner.y - cfg.element.height;
+  switch (cfg.position) {
+    case 'left':
+      x = cfg.inner.x - cfg.element.width;
+      y = cfg.inner.cy - cfg.element.height / 2;
+      break;
+    case 'right':
+      x = cfg.inner.right;
+      y = cfg.inner.cy - cfg.element.height / 2;
+      break;
+    case 'bottom':
+      x = cfg.inner.cx - cfg.element.width / 2;
+      y = cfg.inner.bottom;
+      break;
+  }
+
+  return new Rect(x, y, cfg.element.width, cfg.element.height);
+}
+
+/**
+ * Get opposite position.
+ * @param position - name of position
+ * */
+function getOppositePosition(position: PositionType): PositionType {
+  return ({
+    top: 'bottom',
+    left: 'right',
+    right: 'left',
+    bottom: 'top'
+  }[position] || position) as PositionType;
+}
+
+/**
+ * Update popup and arrow positions to fit by major axis.
+ * @param cfg - popup position config
+ * @param rect - popup position rect
+ * @param arrow - arrow position value
+ * */
+function fitByMajorAxis(cfg: PopupPositionConfig, rect: Rect, arrow: ArrowPositionValue) {
+  if (cfg.behavior !== 'fit') return;
+
+  let isMirrored = false;
+  switch (cfg.position) {
+    case 'bottom':
```

Comment:
DIscussable super cosmetic:
It may look compacter in case we replace the switch with extra conditions. 
But not sure here, so it's definitely rejectable super cosmetic.

Reply:
This is unlikely to give an improvement in this particular place.
But I did manage to do this trick later on this code.